lib/ads/media_tailor_ad_manager.js

```diff
@@ -0,0 +1,860 @@
+   * @private
+   */
+  sendTrackingEvent_(eventType) {
+    let trackingEvent = this.mediaTailorAd_.trackingEvents.find(
+        (event) => event.eventType == eventType);
+    if (!trackingEvent) {
+      trackingEvent = this.mediaTailorAdBreak_.adBreakTrackingEvents.find(
+          (event) => event.eventType == eventType);
+    }
+    if (trackingEvent) {
+      const type = shaka.net.NetworkingEngine.RequestType.ADS;
+      for (const beaconUrl of trackingEvent.beaconUrls) {
+        if (!beaconUrl || beaconUrl == '') {
+          continue;
+        }
+        const request = shaka.net.NetworkingEngine.makeRequest(
+            [beaconUrl],
+            shaka.net.NetworkingEngine.defaultRetryParameters());
+        request.method = 'POST';
+        this.networkingEngine_.request(type, request);
+      }
+    }
+    switch (eventType) {
+      case 'impression':
+        this.onEvent_(
+            new shaka.util.FakeEvent(shaka.ads.AdManager.AD_IMPRESSION));
+        break;
+      case 'start':
+        this.onEvent_(
+            new shaka.util.FakeEvent(shaka.ads.AdManager.AD_STARTED,
+                (new Map()).set('ad', this.ad_)));
+        break;
+      case 'mute':
+        this.onEvent_(
+            new shaka.util.FakeEvent(shaka.ads.AdManager.AD_MUTED));
+        break;
+      case 'unmute':
+        this.onEvent_(
+            new shaka.util.FakeEvent(shaka.ads.AdManager.AD_VOLUME_CHANGED));
+        break;
+      case 'resume':
+        this.onEvent_(
+            new shaka.util.FakeEvent(shaka.ads.AdManager.AD_RESUMED));
+        break;
+      case 'pause':
+        this.onEvent_(
+            new shaka.util.FakeEvent(shaka.ads.AdManager.AD_PAUSED));
+        break;
+      case 'firstQuartile':
+        this.onEvent_(
+            new shaka.util.FakeEvent(shaka.ads.AdManager.AD_FIRST_QUARTILE));
+        break;
+      case 'midpoint':
+        this.onEvent_(
+            new shaka.util.FakeEvent(shaka.ads.AdManager.AD_MIDPOINT));
+        break;
+      case 'thirdQuartile':
+        this.onEvent_(
+            new shaka.util.FakeEvent(shaka.ads.AdManager.AD_THIRD_QUARTILE));
+        break;
+      case 'complete':
+        this.onEvent_(
+            new shaka.util.FakeEvent(shaka.ads.AdManager.AD_COMPLETE));
+        this.onEvent_(
+            new shaka.util.FakeEvent(shaka.ads.AdManager.AD_STOPPED));
+        break;
+      case 'breakStart':
+        this.adContainer_.setAttribute('ad-active', 'true');
+        break;
+      case 'breakEnd':
+        this.adContainer_.removeAttribute('ad-active');
+        break;
+    }
+  }
+
+  /**
+   * @param {string} url
+   * @param {string} assetUri
+   * @param {?number} errorCode
+   * @return {string}
+   * @private
+   */
+  replaceMacros_(url, assetUri, errorCode) {
+    let finalString = url;
+
+    const pattern1 = /\[CACHEBUSTING\]/gi;
+    if (pattern1.test(finalString)) {
+      finalString = finalString.replace(
+          pattern1, this.generateCacheBusting_());
```

Comment:
None of these replace macro functions seem to use the state of this manager, so they could all be static, right?